src/computability/language.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson.
+-/
+import data.finset.basic
+
+/-!
+# Languages
+This file contains the definition and operations on formal languages over an alphabet. Note strings
+are implemented as lists over the alphabet.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- A language is a set of strings over an alphabet. -/
+@[derive [has_union, has_emptyc, has_mem (list α), has_singleton (list α)]]
+def language (α) := set (list α)
+
+namespace language
+
+instance : inhabited (language α) := ⟨∅⟩
+
+instance : has_zero (language α) := ⟨∅⟩
+instance : has_one (language α) := ⟨{[]}⟩
+
+instance : has_add (language α) := ⟨set.union⟩
+instance : has_mul (language α) := ⟨λ l m, (l.prod m).image (λ p, p.1 ++ p.2)⟩
+
+@[simp] lemma zero_def : (0 : language α) = ∅ := rfl
+@[simp] lemma one_def : (1 : language α) = {[]} := rfl
+
+@[simp] lemma add_def (l m : language α) : l + m = l ∪ m := rfl
+@[simp] lemma mul_def (l m : language α) : l * m = (l.prod m).image (λ p, p.1 ++ p.2) := rfl
+
+/-- The star of a language `L` is the set of all strings which can be written by concatenating
+  strings from `L`. -/
+def star (l : language α) : language α :=
+{ x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, ¬(list.empty y) ∧ y ∈ l}
+
+@[simp] lemma star_def (l : language α) :
+  l.star = { x | ∃ S : list (list α), x = S.join ∧ ∀ y ∈ S, ¬(list.empty y) ∧ y ∈ l} := rfl
+
+private lemma mul_assoc (l m n : language α) : (l * m) * n = l * (m * n) :=
+by ext x; simp; tauto {closer := `[subst_vars, simp *]}
```

Comment:
My understanding was that type type of proof was frowned upon, and it should be using `,` instead:
```suggestion
by { ext x, simp, tauto {closer := `[subst_vars, simp *] }
```
(reserving `;` to indicate when there are actually multiple goals).

I didn't test that suggestion above, mind you - maybe there _are_ multiple goals here.